OAuth 2.0 [DOC-735] (#5422)
segmentio/segment-docs: src/connections/oauth.md

---
title: OAuth 2.0
---

> OAuth 2.0 is an online authorization tool that uses tokens to grant access to API resources like Segment’s tracking API. You can use OAuth 2.0 as a security requirement for connections to third-party tools. 

Reviewer: `online authorization tool` --> `online authorization standard`

`tracking` --> `Tracking`

Author: tracking doesn't need to be capitalized

> info ""

> > This feature is currently in pilot. For pilot, Segment doesn’t support OAuth for Web and Mobile sources where an end-user is involved. OAuth 2.0 only supports server to Segment communication, a machine-to-machine workflow. 

Reviewer: > For pilot, Segment doesn’t support OAuth for Web and Mobile sources where an end-user is involved.

please remove, even GA won't support Web & Mobile

I would add "please reference supported sources table"

> ## Permissions 

Reviewer: Suggestion: add a Glossary section to define OAuth app, scope

Depending on your workspace permissions, your access to OAuth apps is limited. 

Segment Role | Permission
------------ | -----------
Workspace Owner | You can view, create, and edit OAuth apps.
Workspace Member | You cannot view, create, or edit OAuth apps.
Source Admin | You can view and edit OAuth apps. <br> You can connect and disconnect OAuth apps. <br> You can enable or disable OAuth enforcement.
Source Read-only | You can only view OAuth apps.
Function Admin | You can view and edit OAuth apps. <br>You can connect and disconnect OAuth apps. <br>You can enable and disable OAuth enforcement.
Function Read-only | You can only view OAuth apps.

## Create an OAuth app

> info ""
> You must have already created workspace in Segment to use OAuth.

To create a new OAuth application:

> 1. Create a private and public key by running the script:

Reviewer: I would not leave this as the first step. In the `configuration settings` table, I would put these commands under `Public key` details. This is just an example of how to generate public/private key pair, user might use their own method.

    ```
    openssl genpkey -algorithm RSA -out private.pem -pkeyopt rsa_keygen_bits:2048
    openssl rsa -in private.pem -pubout -outform PEM -out public.pem
    ```

2. Navigate to **Settings > Workspace** settings and select the Access Management tab. 
3. Select the **OAuth application** tab within the **Access Management** page. 
4. Click **Create OAuth app**.
5. Enter the configuration settings:

    Settings | Details
    -------- | -------
    Application name | The name of the OAuth app.
    Public key | Upload a public key in PEM format to authenticate through the OAuth application. You can upload a second public key after you create the OAuth application. 
    Public key name | Enter a name for your public key. 
    Token expiration period | You can choose between: 1 day, 2 days, 3 days, 1 week, 2 weeks, 3 weeks, 30 days.
    Scope | This specifies what type of access is needed for each API. See the list of [supported scopes](#supported-scopes).
6. Click **Create**. 

Once you create your OAuth app, you can now connect a source to your OAuth app. 

## Connect a source to OAuth
> info ""

> > OAuth only supports server-side sources and does not support. Currently for beta, only http API and node.js and public api and source functions are other api points that are covered 

Reviewer: Incomplete sentence. Can we add a full list of supported sources?

This is initial list:
HTTP Tracking API
Source Functions
Node.js SDK 
Python SDK
Public API - only full admin access

To connect a source to OAuth: 

1. Navigate to **Connections > Sources**.
2. Select the source you want to enable OAuth for. 
3. Go to the **Settings** tab of the source page and select **OAuth app**. 

> 4. Click **Create OAuth app**. 

Reviewer: If an OAuth app is already created, this should be click `Connect OAuth app`

5. Select the OAuth app you want to connect the source to. 
6. Click **Connect**. 
7. Select the toggle to **Enable OAuth** for the source. 

> To disconnect the source from OAuth, select **Disconnect**.

Reviewer: Can we distinguish between connect/disconnect OAuth app vs enable/disable OAuth enforcement.

Reviewer: I would do 2 sections `Connect` and `Enable`

Author: connect and enable are part of the same flow

## Request the access token

To request the access token, run:

```

> ./gentoken.sh -k <private-key.pem> -i <key_id> -a <oauth_app_id> | jq '.access_token'

Reviewer: @stangeorge @srikanth-mohan This is where we should link to the publicly accessible API docs for the token endpoint. I believe we agreed that we didnt want to expose `gentoken.sh` since its an internal manual script

To edit an existing OAuth application: 

1. Navigate to **Settings > Workspace** settings and select the **Access Management** tab. 
2. Select the **OAuth application** tab within the **Access Management** page. 
3. Click the application name of the OAuth application you want to edit. 
4. On the **Overview** tab you can:
    1. Revoke a token
    2. Copy the Application ID and the Public key
    3. Delete the OAuth application
5. Select the **Settings** tab on the right window where you can: 
    1. Edit the **Application name**
    2. Delete a public key
    3. Add a new public key
    4. Change the token expiration period
    5. Edit your scope
6. Click **Save changes**. 

## Delete an OAuth app
> info ""
> In order to delete an OAuth app, you must remove all connected sources from the app. 

To delete an OAuth app: 
1. Navigate to **Settings > Workspace settings** and select the **Access Management** tab. 
2. Select the **OAuth application** tab within the **Access Management** page. 
3. Select the **App name** of the OAuth app you want to delete. 
4. Select **Delete OAuth app**.
5. Enter the name of the OAuth app you want to delete. 
6. Click **Delete OAuth app**. 

> ## Remove a token

Reviewer: Typically referred to as "Revoke". Let's also add brief description of why a user would need to revoke a token.

---
title: OAuth 2.0
hidden: true
---

OAuth 2.0 is an online authorization standard that uses tokens to grant access to API resources like Segment’s tracking API. You can use OAuth 2.0 as a security requirement for connections to third-party tools. 

> info ""

> > This feature is currently in pilot and is governed by Segment’s [First Access and Beta Preview Terms](https://www.twilio.com/en-us/legal/tos){:target="_blank"}.  OAuth 2.0 only supports server to Segment communication, a machine-to-machine workflow. Take a look at the [supported sources](#supported-sources) to see what sources are currently supported.

Reviewer 2: ```suggestion
> This feature is currently in pilot and is governed by Segment’s [First Access and Beta Preview Terms](https://www.twilio.com/en-us/legal/tos){:target="_blank"}.  OAuth 2.0 only supports server to Segment communication, a machine-to-machine workflow. Take a look at the [supported sources](#supported-sources) to see what sources Segment supports.
```

## Permissions 
Depending on your workspace permissions, your access to OAuth apps is limited. 

Segment Role | Permission
------------ | -----------
Workspace Owner | You can view, create, and edit OAuth apps.
Workspace Member | You cannot view, create, or edit OAuth apps.
Source Admin | You can view and edit OAuth apps. <br> You can connect and disconnect OAuth apps. <br> You can enable or disable OAuth enforcement.
Source Read-only | You can only view OAuth apps.
Function Admin | You can view and edit OAuth apps. <br>You can connect and disconnect OAuth apps. <br>You can enable and disable OAuth enforcement.
Function Read-only | You can only view OAuth apps.

## Create an OAuth app

> info ""

> > You must have already created workspace in Segment to use OAuth.

Reviewer 2: ```suggestion
> You must have already created a workspace in Segment to use OAuth.
```

To create a new OAuth application:

> 1. Navigate to **Settings > Workspace** settings and select the Access Management tab. 

Reviewer 2: ```suggestion
1. Navigate to **Settings > Workspace settings**  and select the Access Management tab. 
```

Reviewer 2: @stayseesong should ^this be **Workspace settings**?

2. Select the **OAuth application** tab within the **Access Management** page. 
3. Click **Create OAuth app**.
4. Enter the configuration settings:

    Settings | Details
    -------- | -------
    Application name | The name of the OAuth app.
    Public key | Upload a public key in PEM format to authenticate through the OAuth application. You can upload a second public key after you create the OAuth application. You can create a public key by running the script: `openssl rsa -in private.pem -pubout -outform PEM -out public.pem`
    Public key name | Enter a name for your public key. 
    Token expiration period | You can choose between: 1 day, 2 days, 3 days, 1 week, 2 weeks, 3 weeks, 30 days.

>     Scope | This specifies what type of access is needed for each API. See the list of [supported scopes](#supported-scopes).

Reviewer 2: ```suggestion
    Scope | This specifies what type of access you need for each API. See the list of [supported scopes](#supported-scopes).
```

## Connect a source to OAuth
> info ""
> OAuth only supports server-side sources. See the list of [supported sources](#supported-sources).

To connect a source to OAuth: 

1. Navigate to **Connections > Sources**.
2. Select the source you want to enable OAuth for. 
3. Go to the **Settings** tab of the source page and select **OAuth app**. 
4. Click **Connect OAuth app**. 
5. Select the OAuth app you want to connect the source to. 
6. Click **Connect**. 

To disconnect your source from OAuth, click **Disconnect**. 

## Enable a source to OAuth
Once you've connected your source to OAuth, you can enable it. To enable your source:

1. Navigate to **Connections > Sources** and select your source.
2. Go to the **Settings** tab of the source and select **OAuth app**.
3. Turn the toggle on for **Enable OAuth**. 

To disable your source from OAuth, turn the toggle off for **Enable OAuth**.

<!-- ## Request the access token

To request the access token, run:

```
./gentoken.sh -k <private-key.pem> -i <key_id> -a <oauth_app_id> | jq '.access_token'
```
-->

## Edit an OAuth application
To edit an existing OAuth application: 

> 1. Navigate to **Settings > Workspace** settings and select the **Access Management** tab. 

Reviewer 2: ```suggestion
1. Navigate to **Settings > Workspace settings** and select the **Access Management** tab. 
```

2. Select the **OAuth application** tab within the **Access Management** page. 
3. Click the application name of the OAuth application you want to edit. 
4. On the **Overview** tab you can:
    * Revoke a token
    * Copy the Application ID and the Public key
    * Delete the OAuth application
5. Select the **Settings** tab on the right window where you can: 
    * Edit the **Application name**
    * Delete a public key
    * Add a new public key
    * Change the token expiration period
    * Edit your scope
6. Click **Save changes**. 

## Delete an OAuth app
> info ""

> > In order to delete an OAuth app, you must remove all connected sources from the app. 

Reviewer 2: ```suggestion
> To delete an OAuth app, you must remove all connected sources from the app. 
```
